File: Sum-string.cpp
#include <string>
#include <algorithm>

class Solution {
public:
    std::string addStrings(const std::string& num1, const std::string& num2) {
        int i = (int)num1.size() - 1;
        int j = (int)num2.size() - 1;
        int carry = 0;
        std::string sb;
        while (i >= 0 || j >= 0 || carry != 0) {
            int d1 = (i >= 0) ? num1[i] - '0' : 0;
            int d2 = (j >= 0) ? num2[j] - '0' : 0;
            int sum = d1 + d2 + carry;
            carry = sum / 10;
            sb.push_back((sum % 10) + '0');
            i--;
            j--;
        }
        std::reverse(sb.begin(), sb.end());
        return sb;
    }

    bool check(const std::string& s1, const std::string& s2, const std::string& s3) {
        std::string sum = addStrings(s1, s2);
        if (sum.size() > s3.size()) return false;
        if (sum == s3.substr(0, sum.size())) {
            if (sum.size() == s3.size()) return true;
            return check(s2, sum, s3.substr(sum.size()));
        }
        return false;
    }

    bool isSumString(const std::string& s) {
        for (size_t i = 1; i < s.size(); i++) {
            for (size_t j = 1; i + j < s.size(); j++) {
                std::string s1 = s.substr(0, i);
                std::string s2 = s.substr(i, j);
                std::string s3 = s.substr(i + j);
                if (check(s1, s2, s3)) return true;
            }
        }
        return false;
    }
};

File: Sum-string.java
class Solution {
    public String addStrings(String num1, String num2) {
        int i= num1.length()-1;
        int j = num2.length()-1;
        int carry =0;
        StringBuilder sb = new StringBuilder();
        while(i>=0||j>=0||carry!=0)
        {
            int d1=(i>=0)?num1.charAt(i)-'0':0;
            int d2=(j>=0)?num2.charAt(j)-'0':0;
            int sum = d1+d2+carry;
             carry = sum/10;
            sb.append(sum%10);
            i--;
            j--;
        }
        return sb.reverse().toString();
        
    }
    
    public boolean check(String s1, String s2, String s3){
        String sum = addStrings(s1, s2);
        if(sum.length() > s3.length()) return false;
        if(sum.equals(s3.substring(0, sum.length()))){
            if(sum.length() == s3.length()) return true;
            return check(s2, sum, s3.substring(sum.length()));
        }
        
        return false;
    }
    public boolean isSumString(String s) {
        // ArrayList<Long> arr = new ArrayList<>();
        // return backtrack(new StringBuilder(s), arr);
        // StringBuilder sb = new StringBuilder(s);
        for(int i = 1; i < s.length(); i++){
            for(int j = 1; i + j < s.length(); j++){
                String s1 = s.substring(0, i);
                String s2 = s.substring(i, i + j);
                String s3 = s.substring(j + i);
            
                if(check(s1, s2, s3)) return true;
            }
        }
        
        return false;
    }
}